osmnames/prepare_data/prepare_housenumbers/set_street_ids_by_street_name.sql

```diff
@@ -36,16 +36,17 @@ FROM osm_linestring AS street
 WHERE street.parent_id = housenumber.parent_id
       AND street.normalized_name = housenumber.normalized_street
       AND housenumber.street_id IS NULL
-      AND housenumber.normalized_street != '';
+      AND housenumber.normalized_street != ''; --&
 
 -- set street id by fully matching names within range
 UPDATE osm_housenumber AS housenumber
   SET street_id = COALESCE(street.merged_into, street.osm_id)
 FROM osm_linestring AS street
 WHERE st_dwithin(street.geometry, housenumber.geometry_center, 1000)
+      AND (street.parent_id = housenumber.parent_id) IS NOT TRUE
```

Comment:
> The additional condition is so indirect to account for NULLs.

could you elaborate briefly what you mean by this? (and it maybe worth a comment in the SQL to avoid future confusion).

Reply:
Certainly. The issue is that comparisons with `NULL` evaluate to `NULL`. So here's a rough table how that looks:

```
SELECT NULL = NULL; -- null

SELECT 1 = 1; -- true
SELECT NULL = 1;  -- null

SELECT (1 = 1) IS FALSE;  -- false
SELECT (NULL = 1) IS FALSE;  -- false

SELECT (1 = 1) IS NOT TRUE;  -- false
SELECT (NULL = 1) IS NOT TRUE; -- true
```

The booleans flip for `!=`, but the `NULL`s do not, hence these rows are ignored for the query.

We need this condition to be the inverse of the one in the preceding query (i.e. just `street.parent_id = housenumber.parent_id`). This means we need it to be  `true`  for the following conditions:

- any of the values is `NULL`
- both of the values are `NULL`
- both of the values are `NOT NULL` and also not equal (`!=`)

So:
- if either or both of the values are `NULL` the parens evaluate to `NULL`, which `IS NOT TRUE` ⇒ true
- if both values are present and not equal, the parens evaluate to `FALSE`, which `IS NOT TRUE` ⇒ true
- if the values are equal, the parens are `TRUE`, for which `IS NOT TRUE` ⇒ false

Effectively I am using `NOT` to convert `NULL` into a boolean that I can compare.

A logical equivalent query should be `(street.parent_id != housenumber.parent_id OR street.parent_id IS NULL OR housenumber.parent_id IS NULL)`. This is more straight forward to read, but it hides the relationship with the previous query (i.e. "all conditions are the same, except this one which is the inverse").

Now the challenge is condensing that into a single comment that makes sense. Maybe:

```sql
-- "(some_query) IS NOT TRUE" is the inverse of "some_query" even when some columns are NULL
```

?

Reply:
This trick is used a couple of times already (albeit with `IS NOT FALSE`), but this is the only usage that has a comment:

```sql
-- delete_unusable_entries.sql

-- remove all rows where all names are empty or null values
DELETE FROM osm_polygon WHERE name = '' IS NOT FALSE; --&
DELETE FROM osm_point WHERE name = '' IS NOT FALSE; --&
DELETE FROM osm_linestring WHERE name = '' IS NOT FALSE; --&
```

With that pattern, the comment would be:
```
-- select all rows where values are not equal, or any of them is NULL
-- See https://github.com/OSMNames/OSMNames/pull/225#discussion_r1141441343
```

Reply:
thanks for your explanation for dummies 😄 

> This trick is used a couple of times already...

it's been a long time 😄